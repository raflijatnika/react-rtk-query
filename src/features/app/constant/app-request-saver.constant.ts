export enum REQUEST_SAVER {
  // TODO
  TODO_FETCH_LIST_REQUEST = 'TODO_FETCH_LIST_REQUEST',
  TODO_FETCH_DETAIL_REQUEST = 'TODO_FETCH_DETAIL_REQUEST',
  TODO_CREATE_REQUEST = 'TODO_CREATE_REQUEST',
  TODO_UPDATE_REQUEST = 'TODO_UPDATE_REQUEST',
  TODO_DELETE_REQUEST = 'TODO_DELETE_REQUEST'
}
