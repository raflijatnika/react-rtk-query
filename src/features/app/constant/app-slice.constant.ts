// Interfaces
import { IAppSliceInitialState } from '@/features/app/interfaces/app-slice.interface'

export enum APP_SLICE_LOADING {
  BASE = 'isLoading',
  DETAIL = 'isDetailLoading',
  ACTION = 'isActionLoading',
  DOWNLOAD = 'isDownloadLoading'
}

export const APP_SLICE_INITIAL_STATE: IAppSliceInitialState = {
  app_isInitialized: false,
  app_isConnected: true
}
