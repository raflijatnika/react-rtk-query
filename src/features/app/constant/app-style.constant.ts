export enum APP_COLOR {
  DARK = '#06152B',
  WHITE = '#fff'
}

export enum APP_COLOR_LIGHT {
  PRIMARY = '#3A36DB',
  BACKGROUND = '#F1F4FA'
}
