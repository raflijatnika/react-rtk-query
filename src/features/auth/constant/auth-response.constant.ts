// Interfaces
import {
  IAuthResponseLogin,
  IAuthResponseAuthenticatedUser
} from '@/features/auth/interfaces/auth-response.interface'

export const AUTH_RESPONSE_LOGIN: IAuthResponseLogin = {
  token: 'asdf',
  refreshToken: 'asdf'
}

export const AUTH_RESPONSE_AUTHENTICATED_USER: IAuthResponseAuthenticatedUser =
  {
    id: 'asdf',
    name: 'John Doe',
    email: 'johndoe@gmail.com'
  }
