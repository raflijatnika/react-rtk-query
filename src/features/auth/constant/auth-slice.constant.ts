// Interfaces
import { IAuthSliceInitialState } from '@/features/auth/interfaces/auth-slice.interface'

export const AUTH_SLICE_INITIAL_STATE: IAuthSliceInitialState = {
  auth_authenticatedUser: {
    id: null,
    name: null,
    email: null
  },
  auth_token: {
    token: null,
    refreshToken: null
  }
}
