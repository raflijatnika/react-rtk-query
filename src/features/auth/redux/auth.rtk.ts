// Interfaces
import { IAuthAttrsLogin } from '@/features/auth/interfaces/auth-attrs.interface'
import {
  IAuthResponseToken,
  IAuthResponseAuthenticatedUser
} from '@/features/auth/interfaces/auth-response.interface'

// Rtk
import { emptySplitApi } from '@/features/app/redux/app.rtk'

// Mutations
import { auth_SET_TOKEN } from './auth.slice'

export const authApi = emptySplitApi.injectEndpoints({
  endpoints: builder => ({
    auth_register: builder.mutation<
      IAuthResponseAuthenticatedUser,
      IAuthAttrsLogin
    >({
      query: ({ body }) => ({
        url: '/auth/register',
        method: 'POST',
        body
      })
    }),
    auth_login: builder.mutation<IAuthResponseToken['result'], IAuthAttrsLogin>(
      {
        query: ({ body }) => ({
          url: '/auth/login',
          method: 'POST',
          body
        }),
        transformResponse: (response: IAuthResponseToken) => response.result,
        onQueryStarted: async (payload, { dispatch, queryFulfilled }) => {
          try {
            const response = await queryFulfilled

            dispatch(auth_SET_TOKEN(response.data))
          } catch {
            //
          }
        }
      }
    ),
    auth_me: builder.query<IAuthResponseAuthenticatedUser['result'], void>({
      query: () => ({
        url: `/auth/me`
      })
    })
  }),
  overrideExisting: false
})

export const {
  useAuth_registerMutation,
  useAuth_loginMutation,
  useAuth_meQuery
} = authApi
