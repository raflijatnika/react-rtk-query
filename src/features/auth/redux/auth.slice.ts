// Redux Toolkit
import { createSlice, PayloadAction } from '@reduxjs/toolkit'

// Constant
import { AUTH_SLICE_INITIAL_STATE } from '@/features/auth/constant/auth-slice.constant'

// Interfaces
import {
  IAuthToken,
  IAuthAuthenticatedUser
} from '@/features/auth/interfaces/auth.interface'

const auth = createSlice({
  name: 'auth',
  initialState: AUTH_SLICE_INITIAL_STATE,
  reducers: {
    auth_SET_TOKEN: (
      state,
      { payload: { token, refreshToken } }: PayloadAction<IAuthToken>
    ): void => {
      state.auth_token.token = token
      state.auth_token.refreshToken = refreshToken
    },
    auth_SET_AUTHENTICATED_USER: (
      state,
      { payload }: PayloadAction<IAuthAuthenticatedUser>
    ): void => {
      state.auth_authenticatedUser = payload
    },
    auth_LOGOUT: state => {
      state.auth_token = AUTH_SLICE_INITIAL_STATE.auth_token
      state.auth_authenticatedUser =
        AUTH_SLICE_INITIAL_STATE.auth_authenticatedUser
    }
  }
})

// Mutations
export const { auth_SET_TOKEN, auth_SET_AUTHENTICATED_USER, auth_LOGOUT } =
  auth.actions

export default auth.reducer
