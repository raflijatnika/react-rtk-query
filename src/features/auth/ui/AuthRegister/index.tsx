// React
import { useCallback } from 'react'

// Components
import { StyledCredentialsContainer } from '@/features/auth/components'
import {
  AppBaseButton,
  AppBaseDivider,
  AppBaseFormItem,
  AppBaseLabel,
  AppBaseInput,
  AppBaseCheckbox,
  AppBaseSocialMediaButton
} from '@/features/app/components'

// i18n
import { useTranslation } from 'react-i18next'

// Constant
import { APP_COLOR_LIGHT } from '@/features/app/constant/app-style.constant'

// Antd
import { Form } from 'antd'

// React Router DOM
import { useNavigate } from 'react-router-dom'

// Interfaces
import { IAuthRegisterForm } from '@/features/auth/interfaces/auth.interface'

// Rtk
import { useAuth_registerMutation } from '@/features/auth/redux/auth.rtk'
import { notificationUtils_open } from '@/features/app/utils/notification.utils'

const AuthRegister = () => {
  // Hook
  const { t } = useTranslation()
  const [form] = Form.useForm()
  const navigate = useNavigate()

  // Register
  const [auth_register, { isLoading: isRegisterLoading }] =
    useAuth_registerMutation()

  /**
   * @description Redirect to login
   *
   * @return {void} void
   */
  const onRedirectToLogin = useCallback((): void => {
    navigate('/auth/login')
  }, [navigate])

  /**
   * @description Register handler
   *
   * @param {IAuthRegisterForm} form
   *
   * @return {Promise<void>} Promise<void>
   */
  const onSubmit = useCallback(
    async (form: IAuthRegisterForm): Promise<void> => {
      try {
        const registerResponse = await auth_register({ body: form }).unwrap()

        notificationUtils_open('success', {
          message: registerResponse.message
        })

        navigate('/auth/login')
      } catch (_) {
        //
      }
    },
    [navigate, auth_register]
  )

  return (
    <div data-testid='auth-register'>
      {/* Social Media Button */}
      <AppBaseSocialMediaButton />

      {/* Divider */}
      <div className='mt-4 w-full'>
        <AppBaseDivider />
      </div>

      {/* Form */}
      <Form
        form={form}
        onFinish={onSubmit}
        layout='vertical'
        requiredMark={false}
      >
        {/* Full Name */}
        <AppBaseFormItem
          name='fullName'
          label={t('app.user.fullName')}
          rules={[{ required: true }]}
        >
          <AppBaseInput placeholder={t('app.user.fullName')} />
        </AppBaseFormItem>

        {/* Email */}
        <AppBaseFormItem
          name='email'
          label={t('app.user.email')}
          rules={[{ required: true, type: 'email' }]}
        >
          <AppBaseInput placeholder='example@gmail.com' />
        </AppBaseFormItem>

        {/* Password */}
        <AppBaseFormItem
          name='password'
          label={t('app.user.password')}
          rules={[{ required: true, min: 8, type: 'string' }]}
        >
          <AppBaseInput type='password' placeholder={t('app.user.password')} />
        </AppBaseFormItem>

        {/* Remember Me & Reset Password  */}
        <StyledCredentialsContainer>
          {/* Checkbox Of Remember Me */}
          <AppBaseFormItem
            name='agree'
            valuePropName='checked'
            rules={[
              {
                validator: (_, value) =>
                  value
                    ? Promise.resolve()
                    : Promise.reject(new Error('Should accept agreement'))
              }
            ]}
          >
            <AppBaseCheckbox>
              <AppBaseLabel fontSize={16}>
                {t('auth.privacy')}{' '}
                <AppBaseLabel fontColor={APP_COLOR_LIGHT.PRIMARY} fontSize={16}>
                  {t('auth.termsOfUse').toLowerCase()}{' '}
                </AppBaseLabel>
                {t('app.and').toLowerCase()} {t('app.our').toLowerCase()}{' '}
                <AppBaseLabel fontColor={APP_COLOR_LIGHT.PRIMARY} fontSize={16}>
                  {t('auth.policy').toLowerCase()}
                </AppBaseLabel>
              </AppBaseLabel>
            </AppBaseCheckbox>
          </AppBaseFormItem>
        </StyledCredentialsContainer>

        {/* Register Button */}
        <AppBaseButton
          type='primary'
          htmlType='submit'
          height={50}
          className='mt-6'
          loading={isRegisterLoading}
          block
        >
          {t('auth.createAccount')}
        </AppBaseButton>

        {/* Information if user didn't have account */}
        <div className='flex items-center gap-1 mt-5 justify-center'>
          <AppBaseLabel fontSize={16}>
            {t('auth.alreadyHaveAccount')}?
          </AppBaseLabel>
          <AppBaseLabel
            fontColor={APP_COLOR_LIGHT.PRIMARY}
            className='cursor-pointer'
            onClick={onRedirectToLogin}
            fontSize={16}
            data-testid='login-button'
          >
            {t('auth.login')}
          </AppBaseLabel>
        </div>
      </Form>
    </div>
  )
}

export { AuthRegister }
