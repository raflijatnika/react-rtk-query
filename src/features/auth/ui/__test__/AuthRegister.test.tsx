// Test
import { act, screen } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import 'jest-styled-components'

// UI
import { AuthRegister } from '../AuthRegister'

// Plugins
import { testRedux_renderWithProviders } from '@/plugins/redux/test.redux'

// Constant
import { AUTH_SLICE_INITIAL_STATE } from '@/features/auth/constant/auth-slice.constant'

// React Router DOM
import { useNavigate } from 'react-router-dom'

it('AuthRegister should render', () => {
  testRedux_renderWithProviders(<AuthRegister />, {
    useRouting: true,
    preloadedState: {
      auth: {
        ...AUTH_SLICE_INITIAL_STATE,
        _persist: { version: 1, rehydrated: true }
      }
    }
  })
})

it('AuthRegister should have snapshot', () => {
  const { container } = testRedux_renderWithProviders(<AuthRegister />, {
    useRouting: true,
    preloadedState: {
      auth: {
        ...AUTH_SLICE_INITIAL_STATE,
        _persist: { version: 1, rehydrated: true }
      }
    }
  })

  expect(container.innerHTML).toMatchSnapshot()
})

it('AuthRegister onRedirectToLogin function should defined', async () => {
  await act(() => {
    testRedux_renderWithProviders(<AuthRegister />, {
      useRouting: true
    })
  })

  const user = userEvent.setup()

  await user.click(screen.getByTestId('login-button'))

  const navigate = useNavigate()

  expect(navigate).toHaveBeenCalled()
})

it('AuthRegister onSubmit function should defined', async () => {
  //
})
