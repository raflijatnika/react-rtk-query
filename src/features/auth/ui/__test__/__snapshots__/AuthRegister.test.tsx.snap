// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AuthRegister should have snapshot 1`] = `"<div data-testid=\\"auth-register\\"><div data-testid=\\"wrapper\\" class=\\"sc-iBkjds kcPEiM\\"><button data-testid=\\"google\\" color=\\"#06152B\\" type=\\"button\\" class=\\"ant-btn ant-btn-default sc-dkzDqf jOMXzK\\"><img src=\\"google.png\\" width=\\"24px\\" height=\\"24px\\"><span>app.social.google</span></button><button data-testid=\\"facebook\\" color=\\"#06152B\\" type=\\"button\\" class=\\"ant-btn ant-btn-default sc-dkzDqf jOMXzK\\"><img src=\\"facebook.png\\" width=\\"9px\\" height=\\"18px\\"><span>app.social.facebook</span></button></div><div class=\\"mt-4 w-full\\"><div class=\\"ant-divider ant-divider-horizontal sc-hKMtZM hBbZGS\\" theme=\\"[object Object]\\" role=\\"separator\\"></div></div><form class=\\"ant-form ant-form-vertical ant-form-hide-required-mark\\"><div class=\\"ant-form-item sc-eCYdqJ cFdIic\\"><div theme=\\"[object Object]\\" class=\\"ant-row ant-form-item-row\\"><div class=\\"ant-col ant-form-item-label\\"><label for=\\"fullName\\" class=\\"ant-form-item-required\\" title=\\"app.user.fullName\\">app.user.fullName</label></div><div class=\\"ant-col ant-form-item-control\\"><div class=\\"ant-form-item-control-input\\"><div class=\\"ant-form-item-control-input-content\\"><input placeholder=\\"app.user.fullName\\" id=\\"fullName\\" class=\\"ant-input sc-jSMfEi dcJQax\\" type=\\"text\\" value=\\"\\"></div></div></div></div></div><div class=\\"ant-form-item sc-eCYdqJ cFdIic\\"><div theme=\\"[object Object]\\" class=\\"ant-row ant-form-item-row\\"><div class=\\"ant-col ant-form-item-label\\"><label for=\\"email\\" class=\\"ant-form-item-required\\" title=\\"app.user.email\\">app.user.email</label></div><div class=\\"ant-col ant-form-item-control\\"><div class=\\"ant-form-item-control-input\\"><div class=\\"ant-form-item-control-input-content\\"><input placeholder=\\"example@gmail.com\\" id=\\"email\\" class=\\"ant-input sc-jSMfEi dcJQax\\" type=\\"text\\" value=\\"\\"></div></div></div></div></div><div class=\\"ant-form-item sc-eCYdqJ cFdIic\\"><div theme=\\"[object Object]\\" class=\\"ant-row ant-form-item-row\\"><div class=\\"ant-col ant-form-item-label\\"><label for=\\"password\\" class=\\"ant-form-item-required\\" title=\\"app.user.password\\">app.user.password</label></div><div class=\\"ant-col ant-form-item-control\\"><div class=\\"ant-form-item-control-input\\"><div class=\\"ant-form-item-control-input-content\\"><input type=\\"password\\" placeholder=\\"app.user.password\\" id=\\"password\\" class=\\"ant-input sc-jSMfEi dcJQax\\" value=\\"\\"></div></div></div></div></div><div class=\\"sc-bczRLJ hvLtLW\\"><div class=\\"ant-form-item sc-eCYdqJ cFdIic\\"><div theme=\\"[object Object]\\" class=\\"ant-row ant-form-item-row\\"><div class=\\"ant-col ant-form-item-control\\"><div class=\\"ant-form-item-control-input\\"><div class=\\"ant-form-item-control-input-content\\"><label class=\\"ant-checkbox-wrapper ant-checkbox-wrapper-in-form-item sc-gKXOVf\\"><span class=\\"ant-checkbox\\"><input id=\\"agree\\" type=\\"checkbox\\" class=\\"ant-checkbox-input\\" value=\\"\\"><span class=\\"ant-checkbox-inner\\"></span></span><span><label font-size=\\"16\\" class=\\"sc-gsnTZi llsaSf\\">auth.privacy <label font-size=\\"16\\" class=\\"sc-gsnTZi gnknbY\\">auth.termsofuse </label>app.and app.our <label font-size=\\"16\\" class=\\"sc-gsnTZi gnknbY\\">auth.policy</label></label></span></label></div></div></div></div></div></div><button type=\\"submit\\" class=\\"ant-btn ant-btn-primary ant-btn-block sc-dkzDqf leEBRr mt-6\\"><span>auth.createAccount</span></button><div class=\\"flex items-center gap-1 mt-5 justify-center\\"><label font-size=\\"16\\" class=\\"sc-gsnTZi llsaSf\\">auth.alreadyHaveAccount?</label><label class=\\"sc-gsnTZi gnknbY cursor-pointer\\" font-size=\\"16\\" data-testid=\\"login-button\\">auth.login</label></div></form></div>"`;
